File: mueddi++/mueddi/dawg.cc
#include "dawg.hh"
#include "decoder.hh"
#include "encoder.hh"

#include <algorithm>
#include <set>
#include <stdexcept>
#include <assert.h>
#include <string.h>

namespace mueddi
{

class Builder
{
public:
    Builder(bool root_final);
    Builder(const Builder &) = delete;
    Builder &operator=(const Builder &) = delete;

    void build(const TWords &words);

    Dawg dawg;

private:
    using TRegister = std::set<DawgStateRef>;

    void replace_or_register(DawgStateRef state);

    void add_suffix(const DawgStateRef &state, const std::string &suffix);

    TRegister registr;
};

DawgStateRef DawgState::get_child(uint32_t letter)
{
    auto it = children.find(letter);
    if (it == children.end()) {
        return DawgStateRef();
    } else {
        return it->second;
    }
}

DawgStateRef DawgState::last_child()
{
    auto it = children.rbegin();
    if (it == children.rend()) {
        return DawgStateRef();
    } else {
        return it->second;
    }
}

void DawgState::set_last_child(const DawgStateRef &child)
{
    auto it = children.rbegin();
    assert(it != children.rend());
    it->second = child;
}

void DawgState::add_child(uint32_t letter, const DawgStateRef &child)
{
    auto p = children.insert(TChildren::value_type(letter, child));
    assert(p.second);
}

void DawgState::dump(std::ostream &os) const
{
    char buf[5];

    os << ( finl ? 't' : 'f' ) << " {";
    std::string delim(" ");
    for (TChildren::const_iterator it = children.begin(); it != children.end(); ++it) {
        os << delim;
        utf8_encode(buf, it->first);
        os << '\'' << buf << "\': " << it->second;
        delim = ", ";
    }

    os << " }";
}

Dawg::Dawg(bool root_final):
    root(std::make_shared<DawgState>(root_final))
{
}

Dawg::Dawg(const Dawg &other):
    root(other.root)
{
}

Dawg::~Dawg()
{
}

Dawg &Dawg::operator=(const Dawg &other)
{
    root = other.root;
    return *this;
}

bool Dawg::accepts(const std::string &w)
{
    const unsigned char *u = reinterpret_cast<const unsigned char *>(w.c_str());

    DawgStateRef node = root;
    assert(node.get());

    uint32_t state = UTF8_ACCEPT;
    uint32_t codepoint = 0xdeadbeef;
    while (true) {
        while (decode(&state, &codepoint, *u)) {
            ++u;
        }

        if (state != UTF8_ACCEPT) {
            throw std::runtime_error("invalid UTF-8");
        }

        if (!*u) {
            assert(!codepoint);
            return node->is_final();
        }

        node = node->get_child(codepoint);
        if (!node) {
            return false;
        }

        ++u;
    }
}

Dawg::TPrefixState Dawg::track_prefix(const std::string &word)
{
    const unsigned char *start, *u = reinterpret_cast<const unsigned char *>(word.c_str());
    std::string prefix;
    DawgStateRef next_state = root;
    DawgStateRef prev_state = next_state; // for empty word

    uint32_t state = UTF8_ACCEPT;
    uint32_t ch = 0xdeadbeef;
    while (true) {
        start = u;
        while (decode(&state, &ch, *u)) {
            ++u;
        }

        if (state != UTF8_ACCEPT) {
            throw std::runtime_error("word has invalid UTF-8");
        }

        if (!*u) {
            break;
        }

        ++u;

        prev_state = next_state;
        next_state = prev_state->get_child(ch);
        if (!next_state) {
            break;
        } else {
            prefix.append(reinterpret_cast<const char *>(start), reinterpret_cast<const char *>(u));
        }
    }

    return TPrefixState(prefix, prev_state);
}

inline Builder::Builder(bool root_final):
    dawg(root_final)
{
}

void Builder::build(const TWords &words)
{
    for (auto &word: words) {
        Dawg::TPrefixState prefix_state = dawg.track_prefix(word);
        assert(prefix_state.second.get());
        std::string current_suffix(word.begin() + prefix_state.first.size(), word.end());
        if (prefix_state.second->has_children()) {
            replace_or_register(prefix_state.second);
        }

        add_suffix(prefix_state.second, current_suffix);
    }

    replace_or_register(dawg.root);
}

void Builder::replace_or_register(DawgStateRef state)
{
    assert(state.get());
    DawgStateRef child = state->last_child();
    if (!!child && child->has_children()) {
        replace_or_register(child);
    }

    auto it = registr.find(child);
    if (it != registr.end()) {
        state->set_last_child(*it);
    } else {
        registr.insert(child);
    }
}

void Builder::add_suffix(const DawgStateRef &state, const std::string &suffix)
{
    assert(state.get());
    const unsigned char *u = reinterpret_cast<const unsigned char *>(suffix.c_str());

    DawgStateRef prev_state = state;
    uint32_t char_state = UTF8_ACCEPT;
    uint32_t ch = 0xdeadbeef;
    while (*u) {
        while (decode(&char_state, &ch, *u)) {
            ++u;
        }

        if (char_state != UTF8_ACCEPT) {
            throw std::runtime_error("suffix has invalid UTF-8");
        }

        ++u;

        DawgStateRef next_state = std::make_shared<DawgState>(!*u);
        prev_state->add_child(ch, next_state);
        registr.insert(next_state);
        prev_state = next_state;
    }
}

Dawg make_dawg_impl(TWords &words)
{
    // UTF-8 can be compared per-character, but for the ordering to be
    // by Unicode codepoints (which this program requires), the
    // character must be unsigned
    std::sort(words.begin(), words.end(),
              [](const std::string &a, const std::string &b) {
                  return strcmp(a.c_str(), b.c_str()) < 0;
              });

    Builder builder(words.empty() || words[0].empty());
    builder.build(words);
    return builder.dawg;
}

}

File: mueddir/src/dawg.rs
//! # Deterministic acyclic finite state automaton for word sets
//!
//! AKA Directed acyclic word graph. The algorithm is from Incremental
//! Construction of Minimal Acyclic Finite-State Automata by Jan
//! Daciuk, Stoyan Mihov, Bruce W. Watson and Richard E. Watson.

use std::cell::RefCell;
use std::cmp::Ordering;
use std::collections::{BTreeMap, BTreeSet};
use std::fmt;
use std::rc::Rc;

pub type DawgStateRef = Rc<RefCell<DawgState>>;

pub type Words = Vec<String>;

pub struct DawgState {
    finl: bool,
    children: BTreeMap<char, DawgStateRef>
}

#[derive(Debug)]
pub struct Dawg {
    root: DawgStateRef
}

struct Builder {
    dawg: Dawg,
    register: BTreeSet<DawgStateRef>
}

impl DawgState {
    fn new(finl: bool) -> Self {
        DawgState {
            finl: finl,
            children: BTreeMap::new()
        }
    }

    pub fn get_children(&self) -> &BTreeMap<char, DawgStateRef> {
        &(self.children)
    }

    pub fn is_final(&self) -> bool {
        self.finl
    }

    fn has_children(&self) -> bool {
        !self.children.is_empty()
    }

    fn get_child(&self, letter: char) -> Option<DawgStateRef> {
        match self.children.get(&letter) {
            Some(s) => Some(Rc::clone(&s)),
            None => None
        }
    }

    fn last_child(&self) -> Option<DawgStateRef> {
        if self.children.is_empty() {
            None
        } else {
            let last = self.children.iter().next_back().unwrap().1;
            Some(Rc::clone(&last))
        }
    }

    fn set_last_child(&mut self, child: DawgStateRef) {
        assert!(!self.children.is_empty());
        *self.children.iter_mut().next_back().unwrap().1 = child;
    }

    fn add_child(&mut self, letter: char, child: DawgStateRef) {
        if let Some(_old) = self.children.insert(letter, child) {
            panic!("child for {} already exists", letter);
        }
    }

    fn compare(&self, other: &Self) -> std::cmp::Ordering {
        if self.finl != other.finl {
            if self.finl {
                return Ordering::Greater;
            } else {
                return Ordering::Less;
            }
        }

        let k = self.children.keys().len();
        let l = other.children.keys().len();
        if k != l {
            if k > l {
                return Ordering::Greater;
            } else {
                return Ordering::Less;
            }
        }

        for (sit, oit) in self.children.iter().zip(other.children.iter()) {
            if sit.0 != oit.0 {
                if sit.0 > oit.0 {
                    return Ordering::Greater;
                } else {
                    return Ordering::Less;
                }
            }

            let sp = sit.1.as_ptr();
            let op = oit.1.as_ptr();
            if sp != op {
                if sp > op {
                    return Ordering::Greater;
                } else {
                    return Ordering::Less;
                }
            }
        }

        return Ordering::Equal;
    }
}

impl fmt::Debug for DawgState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", if self.finl { 't' } else { 'f' })?;
        write!(f, "{}", " {")?;

        let mut delim = " ";
        for (k, v) in self.children.iter() {
            write!(f, "{}\'{}\': ", delim, k)?;
            v.fmt(f)?;
            delim = ", ";
        }

        write!(f, "{}", " }")
    }
}

impl PartialEq for DawgState {
    fn eq(&self, other: &Self) -> bool {
        if self.finl != other.finl {
            return false
        }

        if self.children.keys().len() != other.children.keys().len() {
            return false;
        }

        for (sit, oit) in self.children.iter().zip(other.children.iter()) {
            if sit.0 != oit.0 {
                return false;
            }

            if sit.1.as_ptr() != oit.1.as_ptr() {
                return false;
            }
        }

        return true
    }
}

impl Eq for DawgState {}

impl PartialOrd for DawgState {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.compare(&other))
    }
}

impl Ord for DawgState {
    fn cmp(&self, other: &Self) -> Ordering {
        self.compare(other)
    }
}

impl Dawg {
    fn new(root_final: bool) -> Self {
        Dawg {
            root: Rc::new(RefCell::new(DawgState::new(root_final)))
        }
    }

    pub fn get_root(&self) -> DawgStateRef {
        Rc::clone(&(self.root))
    }

    pub fn accepts(&self, w: &str) -> bool {
        let mut state = Rc::clone(&(self.root));
        for ch in w.chars() {
            let child = state.borrow().get_child(ch);
            match child {
                Some(s) => { state = s; },
                None => { return false; }
            }
        }

        let last = state.borrow();
        last.finl
    }

    fn track_prefix(&self, word: &str) -> ( String, DawgStateRef ) {
        let mut prefix = String::new();
        let mut next_state = Rc::clone(&(self.root));
        let mut prev_state = Rc::clone(&next_state);

        for ch in word.chars() {
            prev_state = next_state;
            match prev_state.borrow().get_child(ch) {
                Some(s) => {
                    next_state = s;
                    prefix.push(ch);
                },
                None => {
                    break;
                }
            }
        }

        (prefix, prev_state)
    }
}

impl Builder {
    fn new(root_final: bool) -> Self {
        Builder {
            dawg: Dawg::new(root_final),
            register: BTreeSet::new()
        }
    }

    fn build(&mut self, words: &Vec<String>) {
        for word in words {
            let (common_prefix, last_state) = self.dawg.track_prefix(word);
            let current_suffix = &word[common_prefix.len()..];
            if last_state.borrow().has_children() {
                self.replace_or_register(&last_state);
            }

            self.add_suffix(&last_state, current_suffix);
        }

        let root = Rc::clone(&(self.dawg.root));
        self.replace_or_register(&root);
    }

    fn replace_or_register(&mut self, state: &DawgStateRef) {
        let last = state.borrow().last_child();
        if let Some(child) = last {
            if child.borrow().has_children() {
                self.replace_or_register(&child);
            }

            if let Some(q) = self.register.get(&child) {
                state.borrow_mut().set_last_child(Rc::clone(&q));
            } else {
                self.register.insert(child);
            }
        }
    }

    fn add_suffix(&mut self, state: &DawgStateRef, suffix: &str) {
        let mut prev_state = Rc::clone(state);
        let mut iter = suffix.chars().peekable();
        while let Some(ch) = iter.next() {
            let finl = match iter.peek() {
                Some(_c) => false,
                None => true
            };

            let next_state = Rc::new(RefCell::new(DawgState::new(finl)));
            prev_state.borrow_mut().add_child(ch, Rc::clone(&next_state));
            self.register.insert(Rc::clone(&next_state));
            prev_state = next_state;
        }
    }
}

pub fn make_dawg_impl(words: &mut Words) -> Dawg {
    words.sort();

    let mut builder = Builder::new(words.is_empty() || words[0].is_empty());
    builder.build(words);
    builder.dawg
}
